### DISCUSSION TITLE
how to obtain p-value, correlation and RMSE in CIBERSORT

### DISCUSSION THREAD

**Discussion by Pisto94** [omnideconv/immunedeconv#135] [Q&A] [answered]
Greetings to everyone,
First of all, congratulations for this package. It's absolute amazing and I use it a lot of times.
However, I was wondering how to obtain in CIBERSORT p.value, RMSE and correlation.

I already set LM22 path and CIBERSORT.R in my script with:
set_cibersort_binary("path/CIBERSORT.R")
set_cibersort_mat("path/LM22.txt")

The function I am using is the following:
immunedeconv::deconvolute_cibersort(df_cibersort, arrays = TRUE, absolute = FALSE)
df_cibersort is a df with gene_symbol in rows and samples in column, with obviously values in every cell.

In this way I obtain CIBERSORT values for every immune-cell-subtype, but how can I obtain pvalue, RMSE and correlation?

https://github.com/omnideconv/immunedeconv/blob/master/R/immune_deconvolution_methods.R
https://liulab-dfci.github.io/RIMA/Infiltration.html (in this example, section 6.1, the function is CIBERSORT and so it's quite different i think).

Thank you in advance to everyone

**Comment by grst** [comment, marked as answer]
Hi @Pisto94,

thanks for your questions. Unfortunately, getting RMSE and p-value is currently not possible with immunedeconv. The goal was to provide a unified interface for multiple methods and CIBERSORT is the only method providing these measures. 

If you need these measures, you'll have to revert to running the original CIBERSORT function that's defined in the `CIBERSORT.R` script.

**Reply by Pisto94** [reply]
Hi grst, thank you so much for your answer.

Unfortunately, I am encountering problems with the original CIBERSORT function:

CIBERSORT(
  sig_matrix = lm22,
  mixture_file,
  perm,
  QN = TRUE,
  absolute,
  abs_method = "sig.score"
)

I have run all functions in CIBERSORT.R file.
In sig_matrix I am using path where I save my LM22.txt matrix, and it's all ok I think.
The problem is mixture_file. It is written to use a matrix with row 1 = sample labels; column 1 = gene symbols; no missing values
But when I insert my matrix R aborts the session and I have to restart.

###
here in case there is my simply code and an example of my mixture_file:

CIBERSORT(sig_matrix = "path/to/LM22.txt", mixture_file = "path/to/file_prova.txt", perm = 100)

example of file_prova.txt (matrix)
"gene_symbol" "A853.CEL" "A855.CEL" "A856.CEL" [...]
"A1BG" 6.84713407497099 6.93845030287705 6.31201614769407
"NAT2" 4.26208649801138 4.36933923269729 3.69018121542718
"ADA" 4.61261562758293 4.30772790254146 3.35838693795443 
"CDH2" 4.6278265318482 5.07338794651486 4.51319758606573 
[...]

Any ideas? Is the error in "gene_symbol"? I am trying both with and without but nothing change.

Thank you in advance for time and consideration

**Reply by grst** [reply]
the CIBERSORT call looks alright in principle. 
The mixture file should be tab separated as far as I know (it looks space separated in your sample). 

Any error message you are getting? 

For reference, this is the code immunedeconv uses for calling cibersort:
https://github.com/omnideconv/immunedeconv/blob/28fce560cfdd8a26adac7bba2cb8c6ae9b83c3bb/R/immune_deconvolution_methods.R#L247-L273

**Reply by Pisto94** [reply]
Hi grst, thank you for your support, your time and replies.

I managed to solve the problem.
immunedeconv::deconvolute_cibersort generates itself a table of results in R console, which consists in immune populations score in every sample without pvalue or anything else.
At the same time it generates in my current directory a txt table called "cibersort_results.txt" which includes transposed results and what I was looking for (pvalue, correlation and RMSE).
The pvalue in this case is always 9999 due to a "hidden parameter" in immunedeconv::deconvolute_cibersort (parameter "..." connecting to CIBERSORT function and "perm" parameter. If this parameter is not specified, it is setted to 0 and pvalue results to be 9999.

So to obtain a correct result a had to recover my first comment and use the function in the following way
immunedeconv::deconvolute_cibersort(df_cibersort, arrays = TRUE, absolute = FALSE, perm = 1000)
Then in my current directory I found the cibersort results table with what I was looking for. 

Thank you, this discussion can be closed.
